keps/sig-apps/20190716-cronjob-new-concurrency-policy-catchup.md

```diff
@@ -0,0 +1,175 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives](#alternatives)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document proposes a plan to add a new concurrency policy `CatchUp` to Cronjob,
+such that missed jobs are not skipped and will be scheduled in sequential order.
+Additionally, two derived minor improvements are proposed to better support this feature: 
+(1) Expose the original planned run time to Jobs.
+(2) Only conduct the "maximum 100 missed schedules" check while concurrency policy is set to `Allow`.
+
+## Motivation
+
+Currently if any schedule misses, cronjob controller will skipped it immediately no matter what concurrency policy is in use.
+This behavior, however, does not fit for cronjobs that need to know the exact time range of the data they are going to process.
+For example, I have a cronjob that runs hourly and only processes the logs within the previous hour.
+Say the job now takes 3 hours to run due to a bug, then many jobs will be skipped sporadically,
+and users have to manually find out and backfill the missing pieces in the middle.
+
+In fact this type of stateful behavior is yet supported in Kubernetes at the moment,
+and it causes us not able to migrate all of our cronjobs to Kubernetes.
+Since it is a pretty common practice (e.g. [Airflow](https://airflow.apache.org/scheduler.html#backfill-and-catchup)),
+we would like to enable it with minimum intrusion of the system. Therefore, the original planned run time approach is proposed.
```

Comment:
That is fair, however, Airflow is none-trivial to set up and could be an overkill in some scenarios. I think it would be nice to have an out-of-box solution that provides with just enough building blocks to work with.

Reply:
I tend to agree with Matt here. CronJob is not intended to be a DAG scheduler or workflow orchestrator. We already have a kubernetes native workflow scheduler in the community [Argo](https://github.com/argoproj/argo), and Airflow on Kubernetes is also a project sponsored by the Apache community. I also have concerns about the technical feasibility of counting all completions accurately and successfully backfilling in order, (CronJob is just not designed to do that), but I'll defer to @soltysh and @mortent on that point.

Reply:
Thanks for the feedback and I think your point is valid, but I would like to throw a bit more thoughts from our side.

First of all the `CatchUp` behavior is by no means a DAG scheduler, because it doesn't introduce any dependency between different cronjobs.

Secondly, it might be a little overstated to say it introduces the concept of a "workflow orchestrator", and I would argue on two perspectives:

- Concurrency policy `Replace` isn't a crontab property either, in fact nor `Forbid` does.
- In my opinion, it's more of an "anacron" like behavior rather than an "workflow orchestrator", and it doesn't break the simplicity of Kubernetes CronJob at all.

From an implementation perspective, a simple solution could be use the first unmet schedule time instead of the latest. I'm not sure about if I missed any thing major on "technical feasibility" of the `CatchUp` behavior. Below is an example code I'd imaging:

```
starts := []time.Time{}
...
for t := sched.Next(earliestTime); !t.After(now); t = sched.Next(t) {
    if sj.Spec.ConcurrencyPolicy == v1beta1.AllowConcurrent {
	starts = append(starts, t)
	// An object might miss several starts. For example, if
	// controller gets wedged on friday at 5:01pm when everyone has
	// gone home, and someone comes in on tuesday AM and discovers
	// the problem and restarts the controller, then all the hourly
	// jobs, more than 80 of them for one hourly scheduledJob, should
	// all start running with no further intervention (if the scheduledJob
	// allows concurrency and late starts).
	// However, if there is a bug somewhere, or incorrect clock
	// on controller's server or apiservers (for setting creationTimestamp)
	// then there could be so many missed start times (it could be off
	// by decades or more), that it would eat up all the CPU and memory
	// of this controller. In that case, we want to not try to list
	// all the missed start times.
	// I've somewhat arbitrarily picked 100, as more than 80,
	// but less than "lots".
	if len(starts) > 100 {
		// We can't get the most recent times so just return an empty slice
		return []time.Time{}, fmt.Errorf("too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew")
	}
  }
  if sj.Spec.ConcurrencyPolicy == v1beta1.ForbidConcurrent || sj.Spec.ConcurrencyPolicy == v1alpha1.ReplaceConcurrent {
	if len(starts) == 0 {
		starts = append(starts, t)
	} else {
		starts[0] = t
	}
  }
  if sj.Spec.ConcurrencyPolicy == v1beta1.CatchUpConcurrent {
	starts = append(starts, t)
	break
  }
}
```

The last point I would like to make is that why do we want to force user to use Argo or Airflow, if an out-of-box solution that provides with just enough building blocks to work with is viable.

Reply:
1. I called out Airflow and Agro because it was called out above that back filling missed jobs is "pretty common" functionality. Afaik it is common functionality in workflow orchestrators, but not particularly common in cron schedulers. anacron doesn't actually do backfill in this way and it only runs on daily schedules.

1. From a technical standpoint, my main concern is that the proposal could create conditions where the CronJob in never able to catch up to the current head. i.e. It continuously attempts to complete its backfill and is never able to catch up to its user declared execution schedule.

1. The current direction from sig-architecture is to try to keep the core ("built in") types stable and functional, and to allow extensions to existing functionality using CRDs/custom controllers or API extension servers. If there is enough demand from the community to support this, I think we should at least consider doing it as a SIG sponsored project that lives out of tree (e.g. AnacronJob). However, given that Argo and Airflow exist in the community already, I'm not sure if this is solving a common problem faced by a large number of users (If I'm wrong on this and there is a strong demand for this feature hopefully offering this KEP will demonstrate it via community interest).

Reply:
For point 2, I think `Replace` might have the same or worse issue, because it can potentially always kill the old running job and not able to finish the current one.

For point 3, I have to I agree with you. It's just a bit sad for us because we do not see it make the core less stable or non-functional.

Again, thanks for the feedback.

Reply:
> In my opinion, it's more of an "anacron" like behavior rather than an "workflow orchestrator", and it doesn't break the simplicity of Kubernetes CronJob at all. If `CatchUp` would be considered as a workflow orchestrator, then probably `Replace` and `Forbid` should also be classified the same way.

I'd like to note that anacron never tries to catch up with more than one missed schedule, similar to [how the Kubernetes cronjob controller currently works](https://github.com/kubernetes/kubernetes/blob/a1f4c2ed21ea008c59934a9c6a06fac3da72f963/pkg/controller/cronjob/cronjob_controller.go#L275-L282). That makes its design far simpler than what is proposed in this KEP.